### ISSUE TITLE
Full support for brace shapes

### ISSUE THREAD

**Issue by Pryadka** [rism-ch/verovio#1457]
Having braces corresponding current font is still actual. I can provide 2 solution for supporting brace glyph.

1. Create some .xml files for different brace heights with different names, for instance E0001-brace, E0002-brace …

2. Read data from corresponding glyph file (E000-brace) and extract path. Apply for every point of this pass a transformation with keep the same coordinates for points at the beginning, the end and in the middle, and stretch coordinates in between. The transformation can be non-linear function or some smooth function. The width of unstretchable intervals should be found in experimental way. Horizontal coordinates of the pass can be kept as is or stretched depending on a system height.

@lpugin  What do you think about it.

**Comment by lpugin** [comment]
I would suggest to do the following:
- add an option braceGlyph (bool) in the general layout group. The option will be false by default.
- when true, render using the E000-brace SMuFL.
I think when using a glyph it will be OK to have a linear vertical scaling. We still need to handle this. What needs to be done it to add a `double FontInfo::m_verticalScaling` and double FontInfo::m_horizontalScaling` member and related setter and getter methods. Then add two corresponding parameters to `Doc::GetDrawingSmuflFont` with a default value 1.0. The we can modify the scaling when setting via `DC::SetFont`. In DC::DrawMusicText (and inherited classes), look a the scaling of the font and apply it if not 1.0.
I will add the E000-brace to the Verovio font. OK?

**Comment by rettinghaus** [comment]
The [SMuFL implementation notes](https://w3c.github.io/smufl/gitbook/tables/staff-brackets-and-dividers.html) say:

> The brace glyph should have a height of 1em, i.e. the height of a single five-line stave, and should be scaled proportionally (i.e. in both dimensions, not only in the vertical dimension) in a scoring application to the appropriate height of the two or more staves it encompasses.

**Comment by lpugin** [comment]
@rettinghaus I do not think this is going to look acceptable, particularly if you have more than two staves. See https://github.com/rism-ch/verovio/pull/1349#issuecomment-602540982 This is why we need to scale the height and the width with a different proportion. (SMuFL leaves us free to use the glyphs as we want.)

**Comment by rettinghaus** [comment]
@lpugin Yes, it will not look acceptable with multiple staves, or if they are stretched to wide apart. But SMuFL doesn't exactly allow you to squeeze them like you want. 

To continue the SMuFL comment:

> Font designers may choose to include a number of alternative brace glyphs designed to accommodate larger distances, to avoid the standard brace glyph becoming too wide and bold at larger sizes. (Bravura, for example, includes four stylistic alternates for brace, designed to encompass numbers of staves from one up to 10 or more.)

And it shows __Recommended stylistic alternates__.

**Comment by lpugin** [comment]
I don't think this feature of SMuFL was very well thought. When scaling proportionally and with alternate glyphs for different sizes, you can end up with a brace on three staves that is wider than a brace on four staves - for example. This will look quite weird in my opinion. So we may choose (as stated) not to include alternate glyphs and simply scale them in a way that looks acceptable.

**Comment by msikalo** [comment]
@lpugin   So, shall we consider this approach as approved
-----------------------------------------------------
add an option braceGlyph (bool) in the general layout group. The option will be false by default.
when true, render using the E000-brace SMuFL.
I think when using a glyph it will be OK to have a linear vertical scaling. We still need to handle this. What needs to be done it to add a double FontInfo::m_verticalScaling and double FontInfo::m_horizontalScaling member and related setter and getter methods. They will be 1.0 by default. Then add two corresponding parameters to Doc::GetDrawingSmuflFont with a default value 1.0. Then we can modify the scaling when setting the font via DC::SetFont. In DC::DrawMusicText (and inherited classes), look a the scaling of the font and apply it if not 1.0.
I will add the E000-brace to the Verovio font.
-----------------------------------------------------

?

**Comment by lpugin** [comment]
Yes, this is a good start. We can still adjust it depending on the visual results we obtain, but nothing here will be useless. I will create a develop-brace branch where you can make PRs.

**Comment by msikalo** [comment]
> Yes, this is a good start. We can still adjust it depending on the visual results we obtain, but nothing here will be useless. I will create a develop-brace branch where you can make PRs.

Thank you!
